### PR TITLE
Remove cluster ID from client config

### PR DESCRIPTION
**Pull request by sergeybykov**
My attempt to remove the need to configure cluster ID on the client side and to receive it from gateway instead. I haven't test it extensively yet. Wanted to share the general idea and get feedback.

### CODE REVIEW

**Comment by sebastianburckhardt** [comment]
Looks like we could make it work this way, but it seems a bit fragile to me: there is a good chance that some piece of code uses the wrong client id or gets messed up by the fact that the client id changes (especially if someone else wants to add some functionality to the OutsideRuntimeClient later on).  

In fact, as written, the code it is not quite right (see UnitTests.GeoClusterTests.MultiClusterRegistrationTests.TwoClusterBattery). The problem is that the response message for GetTypeCodeMap never returns, because the activation address passed in the request message was constructed using the temporary handshake id instead of the final client id.

We could fix this  by performing  an initial handshake using low-level operations on the socket. May be easier to understand also, since it would not require to stick the clusterId into the typemap.

**Comment by sebastianburckhardt** [comment]
Looks like the client already does two separate calls to the silo during StartInternal ( GetTypeCodeMap and GetImplicitStreamSubscriberTable ). Perhaps we could consolidate all this back and forth into a single handshake that also includes the clientId.

**Comment by sebastianburckhardt** [comment]
For example, the silo could simply send a handshake struct containing type code map, client id, and implicit stream subscriber table to the client as the first thing on every opened socket.

BTW, why is there a SiloAddress argument to `GetTypeCodeMap` and `GetImplicitStreamSubscriberTable`? It appears to be ignored.

**Comment by sergeybykov** [comment]
> In fact, as written, the code it is not quite right (see UnitTests.GeoClusterTests.MultiClusterRegistrationTests.TwoClusterBattery). The problem is that the response message for GetTypeCodeMap never returns, because the activation address passed in the request message was constructed using the temporary handshake id instead of the final client id.

I'm not sure I follow. If `GetTypeCodeMap()` doesn't return, `StartInternal()` will throw and client initialization will fail. And I'm not seeing that because BVTs are passing. What am I missing?

> Looks like the client already does two separate calls to the silo during StartInternal ( GetTypeCodeMap and GetImplicitStreamSubscriberTable ). Perhaps we could consolidate all this back and forth into a single handshake that also includes the clientId.

`GetImplicitStreamSubscriberTable()` is called from within `StreamingInitialize()` on https://github.com/sergeybykov/orleans/blob/1800-clientid/src/Orleans/Runtime/OutsideRuntimeClient.cs#L306, which is called after the cluster ID is received and clientId is initialized. So it looks correct to me.

> For example, the silo could simply send a handshake struct containing type code map, client id, and implicit stream subscriber table to the client as the first thing on every opened socket.

I don't think we want client to receive cluster ID and type map from every silo it connects to. Client only initializes once - by calling the first gateway it connects to. With support for heterogeneous silos handling of type maps will change. But even in that case we likely won't rely client getting type maps from every silo, because a) client isn't required to connect to every silo; and b) messages received by a gateway get forwarded to other silos, whether the client is connected to the or not. Either way, client doesn't need to receive cluster ID from every silo it connects to, as it can only connect to a single cluster.

I agree that the approach is a bit hacky. I was trying to avoid a major refactoring of the initialization and handshaking in order not to block the PR much further, and anticipating a major refactoring for heterogeneous coming soon.

**Comment by sergeybykov** [comment]
I'm trying to run `TwoClusterBattery` now to reproduce the issue you mentioned. Had to install Azure Storage Emulator for that.

**Comment by benjaminpetit** [comment]
Sergey, I agree that the client should receive the ClusterId at runtime, but I am not sure that `GrainInterfaceMap` is a good place to pass it. Can't we add a specific call for that?

**Comment by sebastianburckhardt** [comment]
> StartInternal()  will throw and client initialization will fail

yes, this is exactly what I am seeing. 

> And I'm not seeing that because BVTs are passing. What am I missing?

I suppose it means that there aren't enough tests in BVT. What does BVT stand for? Should we add some of the geocluster tests?

> GetImplicitStreamSubscriberTable()  is called after the cluster ID is received and clientId is initialized. So it looks correct to me.

Yes, I did not mean to say it is incorrect. But it is slower than necessary. It looks to me like it may take up to three roundtrips before a client is initialized (first establish socket connection, the call GetTypeMap and wait for response, then call GetImplicitStreamSubscriberTable and wait for response). Perhaps this is not important in practice since we are assuming that clients are initialized only once and then used many times. We could add another such call to get the clusterId, I think that is what @benjaminpetit suggests.

> I don't think we want client to receive cluster ID and type map from every silo it connects to

That makes sense. However, it would be easy and fast though to send at least the cluster ID, and that would avoid the problem I see in the failing testcase. Of course there may be other solutions.

**Comment by jdom** [comment]
I agree this is a little hacky, but at least preserves the programming model we want, which is the client just connects to the cluster by knowing it's DeploymentId (btw, I'm not entirely sure why ClusterID != DeploymentID).
IMO a better approach would be that each cluster gateway (not confuse with client gateway) would act as a NAT, and only convert the ClientID to a GeoClientID if it's being sent via a cluster boundary, and then convert it back to a normal ClientID when a message comes to that cluster. Nevertheless this can have issues if the clientid can ever be in the users hands (not sure if it's possible) so that they might serialize it and pass it as an opaque message that ends up being deserialized in a different cluster.
Still this _allegedly_ better alternative would preserve the same programming model as Sergey's workaround

**Comment by sebastianburckhardt** [comment]
My own preference would be just to fix up Sergey's workaround a bit so it works. I will give it a shot. I don't mind if it is a bit hacky. I agree it is preferrable if users don't have to configure the clientId.

**Comment by sergeybykov** [comment]
@jason-bragg and I looked in to the option of adding retrieval of cluster ID to the gateway socket handshake protocol. Our conclusion is that it wouldn't too difficult but not straightforward either because sync sending of handshake and async receiving from the gateway socket are done by rather disjoined pieces code. We thought that it would complicate the GSI PR even more, with such low level communication changes. The suggestion is to open an issue to reconsider this part as an isolated optimization later.

I feel the same about the "NAT" option. It seems too much to try to tie to the long delayed already GSI PR. Can be explored as a potential refactoring later. There might be issue with that approach, too.

To address @benjaminpetit's suggestion, it should be easy to change this PR to have a separate method for obtaining cluster ID. It would add a extra round-trip, but I don't believe it's a major concern, considering that it would happen only once - when client initializes.

I was going to look deeper at the test failure. @sebastianburckhardt, let's collaborate on that.

To summarize, do you agree with the following plan?
1. Fix the issue that breaks the test
2. Move getting cluster ID to a call separate from `GetTypeCodeMap`
3. Test 
4. Merge

**Comment by sebastianburckhardt** [comment]
Yes, I have been looking into this. I have a fix for the bug (the sending grain in the message is updated so that the return message is not lost). Doing that, I realized that it is quite easy on the client to extract the updated clientId from this return message object itself, so it does not need to be tacked onto the type map. I will submit a PR against your PR so you can see what I mean.

**Comment by sebastianburckhardt** [comment]
Ok, I have the suggested fix at sergeybykov/orleans#4

**Comment by sergeybykov** [comment]
Merge https://github.com/sergeybykov/orleans/pull/4. I like it much better than my original attempt.
